frontendActions/src/main/java/com/virtuslab/gitmachete/frontend/actions/common/BaseResetBranchToRemoteAction.java

```diff
@@ -0,0 +1,188 @@
+        VcsNotifier.getInstance(project).notifyWarning(VCS_LOGGER_TITLE,
+            "Skipping the action because no VCS repository is selected");
+      }
+    } else {
+      LOG.warn("Skipping the action because name of branch to reset is undefined");
+      VcsNotifier.getInstance(project).notifyError(VCS_LOGGER_TITLE,
+          "Internal error occurred. For more information see IDE log file");
+    }
+  }
+
+  protected void doResetToRemoteWithKeep(Project project, GitRepository gitRepository, String branchName,
+      IGitMacheteRepository macheteRepository, AnActionEvent anActionEvent) {
+
+    new Task.Backgroundable(project, TASK_TITLE, /* canBeCancelled */ true) {
+
+      @Override
+      public void run(ProgressIndicator indicator) {
+        LOG.debug(() -> "Resetting '${branchName}' branch");
+        try (AccessToken ignored = DvcsUtil.workingTreeChangeStarted(project, TASK_TITLE)) {
+          GitLineHandler resetHandler = new GitLineHandler(myProject, gitRepository.getRoot(), GitCommand.RESET);
+          resetHandler.addParameters("--keep");
+
+          Option<BaseGitMacheteBranch> branchOption = macheteRepository.getBranchByName(branchName);
+          assert branchOption.isDefined() : "Can't get branch '${branchName}' from Git Machete repository";
+          Option<IGitMacheteRemoteBranch> remoteTrackingBranchOption = branchOption.get().getRemoteTrackingBranch();
+          if (remoteTrackingBranchOption.isDefined()) {
+            resetHandler.addParameters(remoteTrackingBranchOption.get().getPointedCommit().getHash());
+          } else {
+            String message = "Branch '${branchName}' doesn't have remote tracking branch, so cannot be reset";
+            LOG.warn(message);
+            VcsNotifier.getInstance(project).notifyWarning(VCS_LOGGER_TITLE, message);
+          }
+
+          resetHandler.endOptions();
+
+          // Check if branch to reset is not current branch - if isn't then checkout
+          Option<String> currentBranchOption = getCurrentBranchNameIfManaged(anActionEvent);
+          if (currentBranchOption.isEmpty() || !currentBranchOption.get().equals(branchName)) {
+            // Checking out given branch
+            GitBranchUiHandlerImpl uiHandler = new GitBranchUiHandlerImpl(project, Git.getInstance(), indicator);
+            new GitBranchWorker(project, Git.getInstance(), uiHandler)
+                .checkout(branchName, /* detach */ false, List.of(gitRepository));
+
+            // Check again if we are in branch to reset to be sure that checkout was successful
+            // This time we are using git4idea because GitMacheteRepository is immutable and it would return previous branch
+            GitLocalBranch nullableLocalBranch = gitRepository.getCurrentBranch();
+            if (nullableLocalBranch == null || !nullableLocalBranch.getName().equals(branchName)) {
+              LOG.error("Checkout to branch ${branchName} before reset was unsuccessful");
```

Comment:
This sentence sounds weird to me. Since this log is preceded by 
`LOG.debug(() -> "Resetting '${branchName}' branch");`
what about logging the fact that the checkout is needed (right in the place where we know that, so the line 156 of this file) and then, here we could just: 
`LOG.error("Checkout to branch ${branchName} failed");`

Alternatively, the following sounds more ~natural~ better to me:
`LOG.error("Reset-preceding checkout of branch ${branchName} failed");`